Feature: Add Posts controller

stephanie56 opened the pull request:
1. Add the `posts` controller
2. Implement the following endpoints
- `GET /posts`
- `GET /posts/:id`
-  `POST /posts`
- `PUT /posts`
- `DELETE /posts`

Commit:
Set up post controller to handle Get and Post requests

Commit:
Add Post and Delete endpoints for posts controller

Commit:
Add express validator

Commit:
Add Swagger config

ysim commented on src/api/index.js:
```diff
@@ -1,8 +1,9 @@
-const express = require("express");
+const express = require('express');
 
-const { healthRouter } = require('../routes/health/health.router')
```
Just curious, was this code autogenerated by Express?

stephanie56 replied:
That's part of the boilerplate code from the `be-cohort1-sample` :)

ysim replied:
Gotcha :)

ysim commented on .gitignore:
```diff
@@ -1,2 +1,3 @@
 node_modules
 *.log
+data.json
```
Looks like you accidentally committed the `data.json` file before adding this to your `.gitignore` :D

stephanie56 replied:
Yes - but I think this rule is not working 🤔I had unstaged data.json, but it's still added.

ysim replied:
Hm, the pathspec is right, so it should be! Anyway, not a big deal, we can sync up on Slack for some tips on checking stuff into the git staging area. In the meantime, you should be able to remove it using `git rm`!

ysim commented on src/routes/posts/posts.controller.js:
```diff
@@ -0,0 +1,93 @@
+
+const postsData = require('../../db/data.json');
+
+const PATH_TO_DATA = 'src/db/data.json';
```
Is this not referring to the same file as the above const `postsData`?

stephanie56 replied:
It refers to the same file, but the path used in writeFile is an absolute path - tried to use a relative path for the writeFile and it didn't work :(

ysim replied:
Oh, interesting. I see that they've done the same thing in the boilerplate: https://github.com/bridge-school/be-cohort1-sample/blob/2.3/error-handling/src/routes/movies/movies.controller.js#L25 

I wonder if this would help keep the code more DRY? https://nodejs.org/api/fs.html#fs_fs_realpath_path_options_callback

stephanie56 replied:
Sounds great! Going to look into it :)

ysim commented on src/routes/posts/posts.controller.js:
```diff
@@ -0,0 +1,93 @@
+
+  if (!postOrNull) {
+    res.status(404).send('Post not found!');
+  }
```
Nice, I like the fail-fast here.

stephanie56 replied:
Thanks 🎉

ysim commented on src/routes/posts/posts.controller.js:
```diff
@@ -0,0 +1,93 @@
+  };
+
+  try {
+    await await writeFile(
```
Typo?

stephanie56 replied:
Good call!

ysim reviewed (commented):
lgtm! No issues from what I can see, except for the possible `await await` typo. 

For next week, could please provide testing instructions? By that I mean, provide steps in your PR description for how to test your changes (e.g. commands that I should run, output that I should see).

ysim commented on src/routes/posts/posts.router.js:
```diff
@@ -0,0 +1,27 @@
+router.get('/:id', getPostById);
+router.put('', updatePostValidator, validatePost, updatePostById);
+router.post('', createPostValidator, validatePost, createPost);
+router.delete('', deletePostById);
```
Sorry, straggler comment as I just noticed this - your routes for `put` and `delete` seem to be missing params. Ping me on slack if you don't know what I'm talking about!

stephanie56 replied:
Good catch! I took the param from the request body - mainly because the project specs define the posts endpoints  like this https://www.notion.so/Project-Specs-project-and-feedback-groups-4d109441ca1a4e44a7db0a9003d02a6f I was about to confirm with the group to ensure whether we should use path param or follow the specs.

stephanie56 replied:
I think it makes more sense to take the param from the path - so updated the route :)

Commit:
Refactor error handling and update README

Commit:
Update .gitignore

stephanie56 commented on .gitignore:
```diff
@@ -1,2 +1,3 @@
 node_modules
 *.log
+src/db
```
🎉 Thanks @ysim! Used `git rm --cached Filename` - removed the db file :)

stephanie56 commented on README.md:
```diff
@@ -1,3 +1,11 @@
+- Run `npm install` to install packages
+- Run `npm run start` to start the application
+- The server is open at `http://localhost:3000/`
+- Swagger documentation is available at `http://localhost:3000/api-docs`
```
Updated `README` with instructions on how to test - I will update the README again after adding some unit tests.

stephanie56 commented:
@ysim I fixed the `await` typo - and I have added examples to the swagger documentation, should be able to test these endpoints from `http://localhost:3000/api-docs`. Will add test instruction on my next PR as well :)
